tests/test_errors.py

```diff
@@ -0,0 +1,147 @@
+
+    with pytest.raises(
+        TypeError,
+        match="Values must have a dtype of torch.complex64 or torch.complex128",
+    ):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+
+def test_t1_non_real_points() -> None:
+    points = torch.rand((2, 10), dtype=torch.complex128)
+    values = torch.randn(10, dtype=torch.complex128)
+
+    with pytest.raises(
+        TypeError,
+        match="Points must have a dtype of torch.float64 as values has "
+        "a dtype of torch.complex128",
+    ):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+
+def test_t1_mismatch_precision() -> None:
+    points = torch.rand((2, 10), dtype=torch.float32)
+    values = torch.randn(10, dtype=torch.complex128)
+
+    with pytest.raises(
+        TypeError,
+        match="Points must have a dtype of torch.float64 as values has "
+        "a dtype of torch.complex128",
+    ):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+    points = torch.rand((2, 10), dtype=torch.float64)
+    values = torch.randn(10, dtype=torch.complex64)
+
+    with pytest.raises(
+        TypeError,
+        match="Points must have a dtype of torch.float32 as values has "
+        "a dtype of torch.complex64",
+    ):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+
+# sizes
+
+
+def test_t1_wrong_length() -> None:
+    points = torch.rand(10, dtype=torch.float64)
+    values = torch.randn(12, dtype=torch.complex128)
+
+    with pytest.raises(
+        ValueError, match="The same number of points and values must be supplied"
+    ):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+    points = torch.rand((3, 10), dtype=torch.float64)
+
+    with pytest.raises(
+        ValueError, match="The same number of points and values must be supplied"
+    ):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+
+def test_t1_points_4d() -> None:
+    points = torch.rand((4, 10), dtype=torch.float64)
+    values = torch.randn(10, dtype=torch.complex128)
+
+    with pytest.raises(ValueError, match="Points can be at most 3d, got"):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+
+def test_t1_too_many_points_dims() -> None:
+    points = torch.rand((1, 4, 10), dtype=torch.float64)
+    values = torch.randn(10, dtype=torch.complex128)
+
+    with pytest.raises(ValueError, match="The points tensor must be 1d or 2d"):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10))
+
+
+def test_t1_wrong_output_dims() -> None:
+    points = torch.rand((2, 10), dtype=torch.float64)
+    values = torch.randn(10, dtype=torch.complex128)
+
+    with pytest.raises(
+        ValueError, match="output_shape must be of length 2 for 2d NUFFT"
+    ):
+        pytorch_finufft.functional.finufft_type1.apply(points, values, (10, 10, 10))
```

Comment:
intended 3d shape here?

Reply:
Yes, this test is for `len(output_shape) != ndim`.

For completion I'll also add one which is too short